plugin-server/src/worker/ingestion/event-pipeline/cookielessServerHashStep.ts

```diff
@@ -0,0 +1,139 @@
+    const datetime = new Date(timestamp)
+    // use the esperanto locale code to get the day of this timestamp in the timezone in YYYY-MM-DD format
+    const day = datetime.toLocaleDateString('eo', { timeZone: timezone || TIMEZONE_FALLBACK })
+
+    // lookup the salt for this day
+    // TODO
+    return 'salt' + day
+}
+
+export function doHash(
+    timestamp: number,
+    timezone: string | undefined,
+    teamId: number,
+    ip: string,
+    host: string,
+    userAgent: string,
+    nIdentifies: number = 0
+) {
+    const salt = getSaltForDay(timestamp, timezone)
+    const key = siphashDouble.string16_to_key(salt)
+    const rootDomain = getDomain(host) || host
+    // use the 128-bit version of siphash to get the result
+    return siphashDouble.hash_hex(key, `${teamId.toString()}-${ip}-${rootDomain}-${userAgent}-${nIdentifies}`)
+}
+
+export async function cookielessServerHashStep(
+    runner: EventPipelineRunner,
+    event: PluginEvent
+): Promise<[PluginEvent | undefined]> {
+    // if events aren't using this mode, skip all processing
+    if (event.properties?.['$device_id'] !== SENTINEL_COOKIELESS_SERVER_HASH_DISTINCT_ID) {
+        return [event]
+    }
+    // drop some events that aren't valid in this mode
+    if (!event.timestamp) {
+        // TODO log
+        return [undefined]
+    }
+    const sessionId = event.properties['$session_id']
+    if (sessionId !== SENTINEL_COOKIELESS_SERVER_HASH_DISTINCT_ID) {
+        // TODO log
+        return [undefined]
+    }
+    // ensure that the distinct id is also the sentinel value
+    if (
+        (event.event === '$identify' &&
+            event.properties['$anon_distinct_id'] !== SENTINEL_COOKIELESS_SERVER_HASH_DISTINCT_ID) ||
+        (event.event !== '$identify' && event.distinct_id !== SENTINEL_COOKIELESS_SERVER_HASH_DISTINCT_ID)
+    ) {
+        // TODO log
+        return [undefined]
+    }
+
+    // drop events that don't have the necessary properties
+    const userAgent = event.properties['$raw_user_agent']
+    const ip = event.properties['$ip']
+    const host = event.properties['$host']
+    const timezone = event.properties['$timezone']
+    const timestamp = DateTime.fromISO(event.timestamp).toMillis()
+    const teamId = event.team_id
+    if (!userAgent || !ip || !host) {
+        // TODO log
+        return [undefined]
+    }
+
+    // drop events from teams that aren't allowed to use this mode
+    const team = await runner.hub.teamManager.getTeamForEvent(event)
+    if (!team?.cookieless_server_hash_opt_in) {
+        // TODO log
+        return [undefined]
+    }
+
+    const hashValue = doHash(timestamp, timezone, teamId, ip, host, userAgent)
+    event.properties['$device_id'] = hashValue
+
+    // TRICKY: if a user were to log in and out, to avoid collisions, we would want a different hash value, so we store the set of identify event uuids for identifies
+    // ASSUMPTION: all events are processed in order, and are processed exactly once
```

Comment:
this assumption doesn't hold, events are at least once, so we need to be able to deal with the same identify event happening twice